pwndbg/commands/xinfo.py

```diff
@@ -0,0 +1,113 @@
+
+import gdb
+
+import pwndbg.arch
+import pwndbg.color.memory as M
+import pwndbg.commands
+import pwndbg.config
+import pwndbg.memory
+import pwndbg.regs
+import pwndbg.stack
+import pwndbg.vmmap
+import pwndbg.wrappers
+
+parser = argparse.ArgumentParser(description='Shows offsets of the specified address to useful other locations')
+parser.add_argument('address', nargs='?', default='$pc', 
+                    help='Address to inspect')
+
+def print_line(name, addr, first, second, op, width = 20):
+
+    print("{} {} = {} {} {:#x}".format(name.rjust(width), M.get(addr),
+        M.get(first) if type(first) is not str else first.ljust(len(hex(addr))),
+        op, second,))
+
+def xinfo_stack(addr):
+    # If it's a stack address, print offsets to top and bottom of stack, as
+    # well as offsets to current stack and base pointer (if used by debugee)
+
+    vmmap = pwndbg.vmmap.find(addr)
+
+    sp = pwndbg.regs.sp
+    frame = pwndbg.regs[pwndbg.regs.frame]
+    frame_mapping = pwndbg.vmmap.find(frame)
+
+    print_line("Stack Top", addr, vmmap.vaddr, addr - vmmap.vaddr, "+")
+    print_line("Stack End", addr, vmmap.end, vmmap.end - addr, "-")
+    print_line("Stack Pointer", addr, sp, addr - sp, "+")
+
+    if frame_mapping and vmmap.vaddr == frame_mapping.vaddr:
+        print_line("Frame Pointer", addr, frame, frame - addr, "-")
+
+    canary_value = pwndbg.commands.canary.canary_value()[0]
+
+    if canary_value is not None:
+        all_canaries = list(
+            pwndbg.search.search(pwndbg.arch.pack(canary_value), mappings=pwndbg.stack.stacks.values())
+        )
+        follow_canaries = sorted(filter(lambda a: a > addr, all_canaries))
+        if follow_canaries is not None and len(follow_canaries) > 0:
+            nxt = follow_canaries[0]
+            print_line("Next Stack Canary", addr, nxt, nxt - addr, "-")
+
+def xinfo_mmap_file(addr):
+    # If it's an address pointing into a memory mapped file, print offsets
+    # to beginning of file in memory and on disk
+
+    vmmap = pwndbg.vmmap.find(addr)
+
+    file_name = vmmap.objfile
+    objpages = filter(lambda p: p.objfile == file_name, pwndbg.vmmap.get())
+    first = sorted(objpages, key = lambda p: p.vaddr)[0]
+    rva = addr - first.vaddr
+
+    print_line("File (Memory)", addr, first.vaddr, rva, "+")
+
+    segments = pwndbg.wrappers.readelf.get_load_segment_info()
+    
+    for s in pwndbg.wrappers.readelf.get_load_segment_info():
+        if rva >= s["VirtAddr"] and rva <= s["VirtAddr"] + s["MemSiz"]:
+            print_line("File (Disk)", addr, file_name, rva - (s["VirtAddr"] - s["Offset"]), "+")
+
+def xinfo_default(addr):
+    # Just print the distance to the beginning of the mapping
+
+    vmmap = pwndbg.vmmap.find(addr)
+
+    print_line("Mapped Area", addr, vmmap.vaddr, addr - vmmap.vaddr, "+")
+
+
+@pwndbg.commands.ArgparsedCommand(parser)
+@pwndbg.commands.OnlyWhenRunning
+def xinfo(address=None):
+    addr = int(address)
+    addr &= pwndbg.arch.ptrmask
```

Comment:
We probably want to have some custom type for commands that gets expressions as arguments (so we don't double those two lines here), but lets leave it as it is for now.